src/librustc_typeck/outlives/mod.rs

```diff
@@ -23,7 +27,51 @@ pub fn provide(providers: &mut Providers) {
 }
 
 //todo
-fn inferred_outlives_of<'a, 'tcx>(_tcx: TyCtxt<'a, 'tcx, 'tcx>, _def_id: DefId)
+fn inferred_outlives_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
                                   -> Vec<ty::Predicate<'tcx>> {
-    Vec::new()
+    // Assert that this is a local node-id
+    let node_id = tcx.hir.as_local_node_id(def_id).unwrap();
+
+    match tcx.hir.get(node_id) {
+        hir_map::NodeItem(item) => match item.node {
+            hir::ItemStruct(..) => Vec::new(),
+            hir::ItemEnum(..) => Vec::new(),
+            _ => Vec::new(),
+        },
+        _ => Vec::new(),
+    }
+}
+
+#[allow(dead_code)]
+fn inferred_outlives_crate <'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
+                                      -> Rc<CratePredicatesMap<'tcx>> {
+    let predicates = FxHashMap();
+    let empty_predicate = Rc::new(Vec::new());
+
+    let empty = Rc::new(
+        ty::CratePredicatesMap {
+            predicates,
+            empty_predicate,
+        }
+    );
+
+    let predicates = tcx.explicit_predicates_of(def_id);
+    let filtered: Vec<_> = predicates
+        .predicates.iter()
+        .filter(|pred|
+            match *pred {
+                &ty::Predicate::TypeOutlives(..) | &ty::Predicate::RegionOutlives(..) => true,
+                _ => false,
+            }
+        )
+        .collect::<Vec<_>>();
```

Comment:
This seems ok -- the idea is then that, in here, we will look over the structs/enums. We'll have (say) a `DefIdMap` (a kind of hashtable keyed by`DefId`) and, for each struct/enum with def-id S, we'll initialize `map[S]` to be `filtered`. 

Though actually maybe it's better to do it slightly differently. Maybe the `map[S]` should be initialized to an empty vector. Then we walk over the fields of S -- if we find that the struct S has a field F that references some other struct S2:

```rust
struct Foo<'a, T> { // Foo is S
    bar: Bar<'a, T> // Bar is S2
}

struct Bar<'b, U: 'b> { ... }
```

then we will (a) compute the filtered, explicit predicates for `Bar` -- this would include `U: 'b` -- and add in any "extra" predicates for `map[Bar]` (in this case, an empty vector). This is the set of requirements that `Bar` imposes to be well-formed; if we translate those back to `Foo` by applying the substitution `[U => T, 'b => 'a]`, then we get the requirements `T: 'a`, which would get added to `Map[Foo]` as one of the implicits we inferred.

In other words, it makes sense to separate out the filtered explicits since we don't really want to add them to the list of implicits, or else we're just duplicating stuff.

Anyway, this is kind of high-level -- would it be helpful if I spelled it out in more low-level detail? Does what I wrote here make sense to you?

Reply:
high-level is actually good and one of the things that I am still a bit fuzzy on. I am going to layout the code-plan as my next step; which should help. After that the implementation details should become easier and plug-and-play.